Feature - support stream for `ChatCompletion`

baradm100 opened the pull request:
Adding support for stream for `ChatCompletion` in the client and in the API.

## How to use?

### Async
```python
from fastchat import client


async def async_main():
    model_name = "vicuna-7b"

    res = await client.ChatCompletion.acreate(
        model=model_name,
        messages=[{"role": "user", "content": "Tell me a story with more than 1000 words."}],
        temperature=0.0,
        max_tokens=32,
        stream=True,
    )
    async for chunk in res:
        content = chunk.choices[0].delta.content
        if content is not None:
            print(content, end="")

if __name__ == "__main__":
    asyncio.run(async_main())
```

### Sync
```python
from fastchat import client
def sync_main():
    model_name = "vicuna-7b"

    res = client.ChatCompletion.create(
        model=model_name,
        messages=[{"role": "user", "content": "Tell me a story with more than 1000 words."}],
        temperature=0.0,
        max_tokens=32,
        stream=True,
    )
    for chunk in res:
        content = chunk.choices[0].delta.content
        if content is not None:
            print(content, end="")

if __name__ == "__main__":
    sync_main()
```

## Issue
Fix #569

kfatehi commented:
hi @baradm100 i dont understand why this PR has to change so many files. i didn't see this implementation when i set out to implement streaming today (shipped in #873), but comparing my work to yours, it appears you are editing way more files. Why?

baradm100 commented:
Hi @kfatehi !
1. I ran the `format.sh` file, as recommended
2. I added `fastchat/serve/test_stream.py` as per the standard to have a file to test a new flow
3. Using BaseModel for the data to standardize and validate the input/output
4. Added support for the SDK as well (for `create` and `acreate`)
5. All the logic that I added that is not needed in the class I moved for helper files

Regarding the re-formating, I know it's intimidating, and I can revert the reformat.

Hazoom commented on fastchat/serve/api.py:
```diff
@@ -141,18 +153,86 @@ def get_gen_params(
+                        )
+                        i += 1
+
+            # Streaming the finsihg token
```
Small typo

Hazoom reviewed (commented):
Amazing work!

Please note that many files have been changed due to the run of `black`, this should be fine I guess but it's worth checking if the version of `black` was the same like in the project requirements.

merrymercy commented:
@baradm100 Thanks for implementing this and rebasing. I will do some local testing and style change and then merge this soon.

merrymercy commented:
@baradm100 This PR is merged! I made some style changes. If you have better ideas, please submit subsequent PRs!

XReyRobert commented:
is it tested compatible with openai.ChatCompletion module ?
No issue in non streaming mode but can't get messages in streaming...

jstzwj commented:
The stream implementation has some inconsistencies with OpenAI's, and I'm fixing those in PR #818.

kfatehi commented:
> is it tested compatible with openai.ChatCompletion module ?
> 
> No issue in non streaming mode but can't get messages in streaming...

I used curl for my test. Do you mind sharing your test code?
